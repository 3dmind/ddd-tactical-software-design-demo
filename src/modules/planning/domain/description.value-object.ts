import { Guard, Result } from '../../../shared/core';
import { ValueObject } from '../../../shared/domain';

interface DescriptionPropsInterface {
  value: string;
}

export class DescriptionValueObject extends ValueObject<
  DescriptionPropsInterface
> {
  public static minLength = 2;
  public static maxLength = 250;

  private constructor(props: DescriptionPropsInterface) {
    super(props);
  }

  get value(): string {
    return this.props.value;
  }

  public static create(text: string): Result<DescriptionValueObject> {
    const nullGuardResult = Guard.againstNullOrUndefined(text, 'text');

    if (!nullGuardResult.succeeded) {
      return Result.fail<DescriptionValueObject>(nullGuardResult.message);
    }

    const minGuardResult = Guard.againstAtLeast(this.minLength, text);
    const maxGuardResult = Guard.againstAtMost(this.maxLength, text);

    if (!minGuardResult.succeeded) {
      return Result.fail<DescriptionValueObject>(minGuardResult.message);
    }

    if (!maxGuardResult.succeeded) {
      return Result.fail<DescriptionValueObject>(maxGuardResult.message);
    }

    return Result.ok<DescriptionValueObject>(
      new DescriptionValueObject({ value: text }),
    );
  }
}
