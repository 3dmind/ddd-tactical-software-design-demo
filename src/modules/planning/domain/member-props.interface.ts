import { UserId } from '../../users/domain/user-id.entity';

export interface MemberProps {
  userId: UserId;
  createdAt?: Date;
}
