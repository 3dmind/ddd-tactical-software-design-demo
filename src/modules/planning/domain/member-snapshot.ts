import { UserId } from '../../users/domain/user-id.entity';
import { MemberId } from './member-id.entity';
import { MemberProps } from './member-props.interface';

export class MemberSnapshot {
  readonly createdAt: Date;
  readonly memberId: MemberId;
  readonly userId: UserId;

  constructor(props: Readonly<MemberProps>, memberId: MemberId) {
    const { createdAt, userId } = props;
    this.createdAt = createdAt;
    this.memberId = memberId;
    this.userId = userId;

    Object.freeze(this);
  }
}
