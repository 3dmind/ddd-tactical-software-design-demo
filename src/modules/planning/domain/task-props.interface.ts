import { Description } from './description.valueobject';

export interface TaskProps {
  archived: boolean;
  archivedAt: Date;
  createdAt: Date;
  description: Description;
  discarded: boolean;
  discardedAt: Date;
  editedAt: Date;
  resumedAt: Date;
  tickedOff: boolean;
  tickedOffAt: Date;
}
