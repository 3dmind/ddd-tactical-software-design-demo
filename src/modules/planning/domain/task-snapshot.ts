import { DescriptionValueObject } from './description.value-object';
import { TaskIdEntity } from './task-id.entity';
import { TaskPropsInterface } from './task-props.interface';

export class TaskSnapshot {
  readonly archivedAt: Date;
  readonly createdAt: Date;
  readonly description: DescriptionValueObject;
  readonly discardedAt: Date;
  readonly editedAt: Date;
  readonly isArchived: boolean;
  readonly isDiscarded: boolean;
  readonly isTickedOff: boolean;
  readonly resumedAt: Date;
  readonly taskId: TaskIdEntity;
  readonly tickedOffAt: Date;

  constructor(props: Readonly<TaskPropsInterface>, taskId: TaskIdEntity) {
    const {
      archived,
      archivedAt,
      createdAt,
      description,
      discarded,
      discardedAt,
      editedAt,
      resumedAt,
      tickedOff,
      tickedOffAt,
    } = props;
    this.archivedAt = archivedAt;
    this.createdAt = createdAt;
    this.description = description;
    this.discardedAt = discardedAt;
    this.editedAt = editedAt;
    this.isArchived = archived;
    this.isDiscarded = discarded;
    this.isTickedOff = tickedOff;
    this.resumedAt = resumedAt;
    this.taskId = taskId;
    this.tickedOffAt = tickedOffAt;

    Object.freeze(this);
  }
}
