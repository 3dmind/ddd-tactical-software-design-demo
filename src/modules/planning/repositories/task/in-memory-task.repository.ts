import { Injectable } from '@nestjs/common';
import { OwnerId } from '../../domain/owner-id.entity';
import { TaskId } from '../../domain/task-id.entity';
import { Task } from '../../domain/task.entity';
import { MaybeTask, TaskRepository } from './task.repository';

/**
 * In-memory implementation of the task repository.
 * Use for test purposes only!
 * @example
 * Test.createTestingModule({
 *   providers: [{
 *     provide: TaskRepository,
 *     useClass: InMemoryTaskRepository,
 *   }]
 * })
 */
@Injectable()
export class InMemoryTaskRepository extends TaskRepository {
  private tasks = new Map<string, Task>();

  constructor() {
    super();
  }

  public async exists(taskId: TaskId): Promise<boolean> {
    return this.tasks.has(taskId.id.toString());
  }

  public async getAllActiveTasksOfOwnerByOwnerId(
    ownerId: OwnerId,
  ): Promise<Task[]> {
    const tasks = Array.from(this.tasks.values());
    return tasks.filter((task) => {
      return (
        task.ownerId.equals(ownerId) &&
        !task.isArchived() &&
        !task.isDiscarded()
      );
    });
  }

  public async getArchivedTasks(): Promise<Task[]> {
    const tasks = Array.from(this.tasks.values());
    return tasks.filter((task) => {
      return task.isArchived() && !task.isDiscarded();
    });
  }

  public async getTaskOfOwnerByTaskId(
    ownerId: OwnerId,
    taskId: TaskId,
  ): Promise<MaybeTask> {
    const task = Array.from(this.tasks.values()).find((task) => {
      return task.ownerId.equals(ownerId) && task.taskId.equals(taskId);
    });
    const found = !!task === true;

    if (found) {
      return {
        found,
        task,
      };
    } else {
      return {
        found,
      };
    }
  }

  public async getTasks(): Promise<Task[]> {
    return Array.from(this.tasks.values());
  }

  public async save(task: Task): Promise<void> {
    this.tasks.set(task.taskId.id.toString(), task);
  }
}
