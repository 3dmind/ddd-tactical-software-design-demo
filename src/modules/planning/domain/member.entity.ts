import { Guard, Result } from '../../../shared/core';
import { Entity, UniqueEntityId } from '../../../shared/domain';
import { UserId } from '../../users/domain/user-id.entity';
import { MemberId } from './member-id.entity';
import { MemberProps } from './member-props.interface';
import { MemberSnapshot } from './member-snapshot';

export class Member extends Entity<MemberProps> {
  private constructor(props: MemberProps, id?: UniqueEntityId) {
    super(props, id);
  }

  get memberId(): MemberId {
    return MemberId.create(this._id).getValue();
  }

  get userId(): UserId {
    return this.props.userId;
  }

  public static create(
    props: MemberProps,
    id?: UniqueEntityId,
  ): Result<Member> {
    const guardResult = Guard.againstNullOrUndefinedBulk([
      {
        argument: props.userId,
        argumentName: 'userId',
      },
    ]);

    if (!guardResult.succeeded) {
      return Result.fail<Member>(guardResult.message);
    }

    const member = new Member(
      {
        ...props,
        createdAt: props.createdAt ?? new Date(),
      },
      id,
    );
    return Result.ok<Member>(member);
  }

  public createSnapshot(): MemberSnapshot {
    return new MemberSnapshot(this.props, this.memberId);
  }
}
