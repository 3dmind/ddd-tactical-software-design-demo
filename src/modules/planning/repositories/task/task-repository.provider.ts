import { Provider } from '@nestjs/common';
import { PrismaTaskRepository } from './prisma-task.repository';
import { TaskRepository } from './task.repository';

export const TaskRepositoryProvider: Provider = {
  provide: TaskRepository,
  useClass: PrismaTaskRepository,
};
