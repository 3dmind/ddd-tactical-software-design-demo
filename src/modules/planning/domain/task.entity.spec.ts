import * as faker from 'faker';
import * as uuid from 'uuid';
import { TaskEntityBuilder } from '../../../../test/builder/task-entity.builder';
import { UniqueEntityId } from '../../../shared/domain';
import { Description } from './description.valueobject';
import { TaskId } from './task-id.entity';
import { TaskProps } from './task-props.interface';
import { Task } from './task.entity';

jest.mock('uuid');

describe('Task', () => {
  beforeAll(() => {
    uuid.v4.mockReturnValue(faker.random.uuid());
  });

  afterAll(() => {
    uuid.mockRestore();
  });

  describe('Guard task properties', () => {
    it('should guard "description against "null" and "undefined"', () => {
      const propsWithNull = { description: null } as TaskProps;
      const propsWithUndefinedDescription = {} as TaskProps;

      const taskResultNull = Task.create(propsWithNull);
      const taskResultUndefined = Task.create(propsWithUndefinedDescription);

      expect(taskResultNull.isFailure).toBe(true);
      expect(taskResultUndefined.isFailure).toBe(true);
    });

    it('should guard "createdAt" against "null" and "undefined"', () => {
      const description = Description.create(faker.lorem.words(5)).getValue();

      const propsWithNull = {
        description,
        createdAt: null,
      } as TaskProps;

      const propsWithUndefinedCreatedDate = {
        description,
      } as TaskProps;

      const taskResultNull = Task.create(propsWithNull);
      const taskResultUndefined = Task.create(propsWithUndefinedCreatedDate);

      expect(taskResultNull.isFailure).toBe(true);
      expect(taskResultUndefined.isFailure).toBe(true);
    });

    it('should guard "tickedOff" against "null" and "undefined"', () => {
      const description = Description.create(faker.lorem.words(5)).getValue();
      const createdAt = new Date();

      const propsWithNull = {
        description,
        createdAt,
        tickedOff: null,
      } as TaskProps;

      const propsWithUndefinedTickedOff = {
        description,
        createdAt,
      } as TaskProps;

      const taskResultNull = Task.create(propsWithNull);
      const taskResultUndefined = Task.create(propsWithUndefinedTickedOff);

      expect(taskResultNull.isFailure).toBe(true);
      expect(taskResultUndefined.isFailure).toBe(true);
    });

    it('should guard "archived" against "null" and "undefined"', () => {
      const description = Description.create(faker.lorem.words(5)).getValue();
      const createdAt = new Date();
      const tickedOff = false;

      const propsWithNull = {
        description,
        createdAt,
        tickedOff,
        archived: null,
      } as TaskProps;

      const propsWithUndefinedArchived = {
        description,
        createdAt,
        tickedOff,
      } as TaskProps;

      const taskResultNull = Task.create(propsWithNull);
      const taskResultUndefined = Task.create(propsWithUndefinedArchived);

      expect(taskResultNull.isFailure).toBe(true);
      expect(taskResultUndefined.isFailure).toBe(true);
    });

    it('should guard "discarded" against "null" and "undefined"', () => {
      const description = Description.create(faker.lorem.words(5)).getValue();
      const createdAt = new Date();
      const tickedOff = false;
      const archived = false;

      const propsWithNull = {
        description,
        createdAt,
        tickedOff,
        archived,
        discarded: null,
      } as TaskProps;

      const propsWithUndefinedDiscarded = {
        description,
        createdAt,
        tickedOff,
        archived,
      } as TaskProps;

      const taskResultNull = Task.create(propsWithNull);
      const taskResultUndefined = Task.create(propsWithUndefinedDiscarded);

      expect(taskResultNull.isFailure).toBe(true);
      expect(taskResultUndefined.isFailure).toBe(true);
    });
  });

  it('should create task', () => {
    const text = faker.lorem.words(5);
    const entityId = new UniqueEntityId();
    const description = Description.create(text).getValue();

    const taskResult = Task.create(
      {
        archived: false,
        archivedAt: null,
        createdAt: new Date(),
        description,
        discarded: false,
        discardedAt: null,
        editedAt: null,
        resumedAt: null,
        tickedOff: false,
        tickedOffAt: null,
      },
      entityId,
    );
    const task = taskResult.getValue();

    expect(taskResult.isSuccess).toBe(true);
    expect(task.taskId).toBeInstanceOf(TaskId);
    expect(task.taskId.id.equals(entityId)).toBe(true);
  });

  it('should note new task', () => {
    const text = faker.lorem.words(5);
    const description = Description.create(text).getValue();

    const taskResult = Task.note(description);
    const task = taskResult.getValue();

    expect(taskResult.isSuccess).toBe(true);
    expect(task.taskId.id.toValue()).toBeDefined();
  });

  it('should tick off task', () => {
    const task = new TaskEntityBuilder().build();

    task.tickOff();

    expect(task.isTickedOff()).toBe(true);
  });

  it('should resume task', () => {
    expect.assertions(1);
    const task = new TaskEntityBuilder().makeTickedOff().build();

    task.resume();

    expect(task.isTickedOff()).toBe(false);
  });

  it('should archive task', () => {
    expect.assertions(1);
    const task = new TaskEntityBuilder().build();

    task.archive();

    expect(task.isArchived()).toBe(true);
  });

  it('should edit description', () => {
    expect.assertions(1);
    const text = faker.lorem.words(5);
    const task = new TaskEntityBuilder(text).build();
    const newText = faker.lorem.words(5);
    const newDescription = Description.create(newText).getValue();

    task.edit(newDescription);

    expect(task.props.description.equals(newDescription)).toBe(true);
  });

  it('should discard task', () => {
    expect.assertions(1);
    const task = new TaskEntityBuilder().build();

    task.discard();

    expect(task.isDiscarded()).toBe(true);
  });
});
