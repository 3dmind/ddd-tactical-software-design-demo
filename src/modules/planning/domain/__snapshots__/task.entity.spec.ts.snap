// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Task should create snapshot 1`] = `
TaskSnapshot {
  "archivedAt": null,
  "createdAt": 1977-01-01T00:00:00.000Z,
  "description": Description {
    "props": Object {
      "value": "Lorem ipsum",
    },
  },
  "discardedAt": null,
  "editedAt": null,
  "isArchived": false,
  "isDiscarded": false,
  "isTickedOff": false,
  "ownerId": OwnerId {
    "_id": UniqueEntityId {
      "value": "74d21847-3298-4775-aaf8-942fed0f53e7",
    },
    "props": null,
  },
  "resumedAt": null,
  "taskId": TaskId {
    "_id": UniqueEntityId {
      "value": "8bb2a83e-cf7e-46d7-a7b0-49f49d94c460",
    },
    "props": null,
  },
  "tickedOffAt": null,
}
`;
